pkg/syncer/upsync/upsync_cleanup_controller.go

```diff
@@ -0,0 +1,191 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package upsync
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/go-logr/logr"
+	kcpcache "github.com/kcp-dev/apimachinery/v2/pkg/cache"
+	kcpdynamic "github.com/kcp-dev/client-go/dynamic"
+	"github.com/kcp-dev/logicalcluster/v3"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	"github.com/kcp-dev/kcp/pkg/indexers"
+	ddsif "github.com/kcp-dev/kcp/pkg/informer"
+	"github.com/kcp-dev/kcp/pkg/logging"
+	syncerindexers "github.com/kcp-dev/kcp/pkg/syncer/indexers"
+	"github.com/kcp-dev/kcp/pkg/syncer/shared"
+)
+
+const cleanupControllerName = "kcp-resource-upsyncer-cleanup"
+
+// NewUpSyncerCleanupController returns a new controller which will cleanup any upsynced upstream resource
+// if the corresponding downstream resources doesn't exist.
+func NewUpSyncerCleanupController(syncerLogger logr.Logger, syncTargetClusterName logicalcluster.Name,
+	syncTargetName string, syncTargetUID types.UID, syncTargetKey string,
+	upstreamClusterClient kcpdynamic.ClusterInterface,
+	ddsifForUpstreamUpsyncer *ddsif.DiscoveringDynamicSharedInformerFactory,
+	ddsifForDownstream *ddsif.GenericDiscoveringDynamicSharedInformerFactory[cache.SharedIndexInformer, cache.GenericLister, informers.GenericInformer],
+) (*cleanupController, error) {
+	c := &cleanupController{
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), cleanupControllerName),
+		cleanupReconciler: cleanupReconciler{
+			getUpstreamClient: func(clusterName logicalcluster.Name) (dynamic.Interface, error) {
+				return upstreamClusterClient.Cluster(clusterName.Path()), nil
+			},
+			getDownstreamLister: func(gvr schema.GroupVersionResource) (cache.GenericLister, error) {
+				informers, notSynced := ddsifForDownstream.Informers()
+				informer, ok := informers[gvr]
+				if !ok {
+					if shared.ContainsGVR(notSynced, gvr) {
+						return nil, fmt.Errorf("informer for gvr %v not synced in the downstream informer factory", gvr)
+					}
+					return nil, fmt.Errorf("gvr %v should be known in the downstream informer factory", gvr)
+				}
+				return informer.Lister(), nil
+			},
+			listDownstreamNamespacesByLocator: func(jsonLocator string) ([]*unstructured.Unstructured, error) {
+				nsInformer, err := ddsifForDownstream.ForResource(namespaceGVR)
+				if err != nil {
+					return nil, err
+				}
+				return indexers.ByIndex[*unstructured.Unstructured](nsInformer.Informer().GetIndexer(), syncerindexers.ByNamespaceLocatorIndexName, jsonLocator)
```

Comment:
Where is ByNamespaceLocatorIndexName added to the DDSIF?

Reply:
When the downstream ddsif is created: https://github.com/kcp-dev/kcp/pull/2675/files#diff-b043590efbb8609f4eced1377afbe4f51358c5c628fb79c0fdd24e8f9d553551R194